src/coreclr/src/System.Private.CoreLib/src/System/GC.cs

```diff
@@ -653,29 +661,63 @@ internal static void UnregisterMemoryLoadChangeNotification(Action notification)
         /// <summary>
         /// Skips zero-initialization of the array if possible.
         /// If T contains object references, the array is always zero-initialized.
+        /// pinned: true means you want to pin this object that you are allocating
+        /// otherwise it's not pinned.
+        /// pinned requires that T does not contain object references.
         /// </summary>
         [MethodImpl(MethodImplOptions.AggressiveInlining)] // forced to ensure no perf drop for small memory buffers (hot path)
-        internal static T[] AllocateUninitializedArray<T>(int length)
+        public static T[] AllocateUninitializedArray<T>(int length, bool pinned = false)
         {
-            if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
+            if (!pinned)
             {
-                return new T[length];
-            }
+                if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
+                {
+                    return new T[length];
+                }
 
-            // for debug builds we always want to call AllocateNewArray to detect AllocateNewArray bugs
+                // for debug builds we always want to call AllocateNewArray to detect AllocateNewArray bugs
 #if !DEBUG
-            // small arrays are allocated using `new[]` as that is generally faster.
-            if (length < 2048 / Unsafe.SizeOf<T>())
+                // small arrays are allocated using `new[]` as that is generally faster.
+                if (length < 2048 / Unsafe.SizeOf<T>())
+                {
+                    return new T[length];
+                }
+#endif
+            }
+            else if (RuntimeHelpers.IsReferenceOrContainsReferences<T>())
             {
-                return new T[length];
+                ThrowHelper.ThrowInvalidTypeWithPointersNotSupported(typeof(T));
             }
-#endif
+
             // kept outside of the small arrays hot path to have inlining without big size growth
```

Comment:
Should the above else if block then also be moved into the AllocateNewUninitializedArray method, to keep it out of the inlinling path?

Reply:
The shape of this method was tweaked to make the case when we fallback to `T[]` fast. That is only relevant if `pinned == false`. (example: look at use in StringBuilder).
I expect that this entire block will be omitted in such case. That is assuming the method does get inlined, which it should.

---
In reply to: [392757230](https://github.com/dotnet/runtime/pull/33526#discussion_r392757230) [](ancestors = 392757230)